### ISSUE TITLE
Design and implement TikTok metric parser

### ISSUE THREAD

**Issue by un7c0rn** [un7c0rn/stopthevirus#1]
Per Slack discussion and @bernardbaker's insight on rate limiting, updating this to be a client side JS lib rather than server side scraper. Also heard back from Adam Heerwagen from **BuildforCOVID19** who was impressed with the project and reached out to Tik Tok Hackathon partners for guidance going forward. No guarantees on Tik Tok whitelisting so proceeding with client parser is best option.

**Comment by bernardbaker** [comment]
That's great news!!

@un7c0rn have I got everything I need to get started with the Firebase work?

**Comment by un7c0rn** [comment]
@bernardbaker I believe so, you've been added as a direct member to the Firestore DB (invitation should be in your email) and schema is in our backend README. Please let me know if you have any questions.

Thanks!

**Comment by bernardbaker** [comment]
Ok.

I should be good to go then.

I'll read through the game cycles to better grasp how the data model would
be used.

I'll be in touch if I have any questions.

**Comment by bernardbaker** [comment]
Could you add me as an **owner** to the Firebase service. I need to configure the service.

https://console.cloud.google.com/iam-admin/iam?authuser=0&project=stv-game-db-test

**Comment by un7c0rn** [comment]
@bernardbaker Yes happy to add you as an owner but would also like to understand the technical requirement. Firestore is a completely dynamic DB meaning that by simply writing to it you may alter the schema. For example submitting an "Entry" document with field "This field wasn't there before" creates the field. Is there a different kind of configuration change we need here?

Also see https://github.com/un7c0rn/stopthevirus/blob/master/game_engine/firestore.py for reference on tribe, player, team, vote and entry usage examples.

Thanks,
Brandon

**Comment by un7c0rn** [comment]
@bernardbaker Done.

**Comment by bernardbaker** [comment]
@un7c0rn  we're using 🔥 store on the web. And I needed to configure phone sign in for logging in users. Unless you plan on using the user name and password. Phone sign in comes with SMS verification.

Also using the 🔥 store admin SDK is great. But we really want a user to only have access to their data. Or data that should be shared with them. Rules need to be applied to read/writes on the 🔥 store DB.

So I needed to setup an app. Remove unneeded accounts. Check GDPR (US/EU) as people's personal information will be processed, stored and or used in some way. Google Analytics is a third party data processor which is configured in Firebase. So that has to be mentioned in the data sub processor web page.

A data protection officer needs to be assigned (you) and I can be the EU representative. Which has a cavet. We need an organisation through G Suite set up. A registered/incorporated company which will solely govern, coordinate GDPR related matters. I would suggest getting in touch with developer community in slack and ask if anyone is US based if you don't want to do it. The role involves some reading, maybe a registration, a fee, but this isn't a public body or educational institution so it may not have to be registered. And not more than 200+ (UK stat) people are employed. But you definitely need to know what the role entails.

Provide a privacy policy, cookies policy and information about the data processors. I'm registered as self employed with our government in the UK. So I could set up a G Suite 💼 account for this. This doesn't need to be done right away 🧂. But is definitely needed if you present the product to people who may question where you are storing the information, how you are processing it and using it. I can also provide a privacy policy which is GDPR compliant along with a cookie policy.

And I may also need to perform any administrative tasks that might pop up while I work on it tomorrow.

**Comment by bernardbaker** [comment]
@un7c0rn I had a look through the python file. I'll implement this in JavaScript tomorrow.

**Comment by un7c0rn** [comment]
Got it. A few thoughts/clarifications:

1. There's no need for client side auth, sign-in or storing user cookies for MVP. Everything can be authenticated by including the phone number and a checksum in the URL sent to the user. This is only possible because we switched to the SMS model. If the primary interaction point was not SMS then you're correct we'd need client side auth. Take this for example:

(A) A new challenge is posted within a game
(B) A user receives a text with a URL to submit their challenge entry. A user may _only_ submit using this URL.
(C) the URL from (B) is formatted as follows: https://<hostname>.io/challenge-submission?a=[challenge identifier]b=[phone number]c=[server checksum]d=[game identifier]
(D) It is explicitly an error to navigate to our challenge submission route without a valid phone number and checksum in the URL.

The client need only verify that the checksum included in the request for navigating to the challenge submission page matches the checksum sent to the user from the backend. This prevents a user from artificially injecting a submission using someone else's phone number. This paradigm should be sufficient for MVP and elide the need for user sessions altogether. I think a goal here is for people to be able to play this game without signing up for a new account or transitively granting access to any PII.

Let me know if you see any issues with this.

2. Agree regarding GDPR, Privacy Policy and ToS but not sure it pertains to the Firestore test instance. Let's create a research/planning document for GDPR that we can use to inform how we setup the prod instance. We may need a retention policy for phone numbers. The Firestore instance we're working off of now is only intended for integration tests and iterative development.

@bernardbaker 

Sounds good re: JS!

**Comment by bernardbaker** [comment]
The checksum will be checked client side by 🔥 store.

**Comment by bernardbaker** [comment]
What personal user information is stored? People will question this?

**Comment by bernardbaker** [comment]
How do you plan to manage user sessions?

**Comment by bernardbaker** [comment]
> Everything can be authenticated by including the phone number and a checksum in the URL sent to the user.

So the phone number and check sum match and they are granted access?

What is some else copies that URL or shares it? If the check some expires 🧂 what if the actual person using the web app with that number loses internet connection? Or refreshes the page?

**Comment by bernardbaker** [comment]
> in the request for navigating to the challenge submission page matches the checksum sent to the user from the backend

So the game ID has a related checksum? And that game ID is extracted and verifed. In the client side code?

**Comment by bernardbaker** [comment]
> This prevents a user from artificially injecting a submission using someone else's phone number.

The phone number is stored, and it will be considered personal information.

**Comment by un7c0rn** [comment]
> What personal user information is stored? People will question this?

As of right now (1) phone number and (2) tik tok handle. Any other PII we'll need to add to our privacy design doc.

**Comment by un7c0rn** [comment]
> How do you plan to manage user sessions?

I don't think we need the concept of a session here. The session is the SMS thread. The context is maintained using UUID's in the request URLs. We can do client side sessions, but may have to brainstorm tradeoffs a bit more. My intention was to minimize UI requirements as much as possible.

**Comment by bernardbaker** [comment]
> I think a goal here is for people to be able to play this game without signing up for a new account or transitively granting access to any PII.

They won't be signing up for an account. It's a secure way to sign into the app. Verify a user session. Login. Logout.

Correct me if I'm wrong but the player enters their phone number. It's in the URL. This is PII.

My response to the aforementioned issues is...
I'm sure this can be managed, and free legal resources can be accessed through developer communities. The app does store phone numbers. The app should provide a sign in method. You can't rely on people playing by the rules, because it's a game.

**Comment by un7c0rn** [comment]
> > Everything can be authenticated by including the phone number and a checksum in the URL sent to the user.
> 
> So the phone number and check sum match and they are granted access?

Yes iff the checksum matches the expected value from the server, then the submission is valid. 

> 
> What is some else copies that URL or shares it? If the check some expires 🧂 what if the actual person using the web app with that number loses internet connection? Or refreshes the page?

I think URL copies, shares  and refreshes are ok, because all we need to do is perform uniqueness validation prior to submitting the entry to the DB. That said there are some interesting considerations here now that I think about it, with regard to using Firestore in the client with the "session-less" SMS model. We may need to consider using the GCP cloud function for all DB writes. Thanks for bringing this up. Additional notes inline.

**Comment by un7c0rn** [comment]
> > This prevents a user from artificially injecting a submission using someone else's phone number.
> 
> The phone number is stored, and it will be considered personal information.

Agreed.

**Comment by un7c0rn** [comment]
> > in the request for navigating to the challenge submission page matches the checksum sent to the user from the backend
> 
> So the game ID has a related checksum? And that game ID is extracted and verifed. In the client side code?

Yes I propose that you can expect that all links generated by the backend and sent via SMS to users will include a server generated checksum for validating URL authenticity. A checksum must match for a given user.

**Comment by bernardbaker** [comment]
> Agree regarding GDPR, Privacy Policy and ToS but not sure it pertains to the Firestore test instance. Let's create a research/planning document for GDPR that we can use to inform how we setup the prod instance. We may need a retention policy for phone numbers. The Firestore instance we're working off of now is only intended for integration tests and iterative development.

I did mention this before.
> This doesn't need to be done right away 🧂.

**Comment by un7c0rn** [comment]
> > in the request for navigating to the challenge submission page matches the checksum sent to the user from the backend
> 
> So the game ID has a related checksum? And that game ID is extracted and verifed. In the client side code?

Actually, no. You raise a great point and I don't want to perform any client side validation in JS. We have a few options here:

1. Implement a session based model using Firestore auth. This comes at the costs of users needing to go through an additional signup flow step, and from my perspective still doesn't solve the malicious user problem since one can simply hijack the JS session in their browser to perform direct writes to the DB. 

2. Move all DB writes to GCP. We can simply extend the Python Firestore abstraction layer we have now and put the DB behind a Python cloud function. This way all writes **must** pass through server side validation for checksums and uniqueness.

**Comment by bernardbaker** [comment]
> > What personal user information is stored? People will question this?
> 
> As of right now (1) phone number and (2) tik tok handle. Any other PII we'll need to add to our privacy design doc.

I'm not sure but that info could be used to identify an individual.

**Comment by un7c0rn** [comment]
> > I think a goal here is for people to be able to play this game without signing up for a new account or transitively granting access to any PII.
> 
> They won't be signing up for an account. It's a secure way to sign into the app. Verify a user session. Login. Logout.
> 
> Correct me if I'm wrong but the player enters their phone number. It's in the URL. This is PII.
> 
> My response to the aforementioned issues is...
> I'm sure this can be managed, and free legal resources can be accessed through developer communities. The app does store phone numbers. The app should provide a sign in method. You can't rely on people playing by the rules, because it's a game.

Need more information on GDPR and PII here. I don't see how a sign in method changes things. In either case we agree that a phone number is PII, so we probably need a retention and wipeout policy. My question is does that policy require a user sign in. I'm not against implementing sign in if it's the right thing to do here. That said I think it's critical to get an end-to-end testable MVP in place before we draft our ToS. I also think per this discussion we cannot write to the Firestore DB from the client. Interested in your thoughts.

Thanks,
Brandon

**Comment by un7c0rn** [comment]
> > > What personal user information is stored? People will question this?
> > 
> > 
> > As of right now (1) phone number and (2) tik tok handle. Any other PII we'll need to add to our privacy design doc.
> 
> I'm not sure but that info could be used to identify an individual.

Yes agreed, it can.

**Comment by bernardbaker** [comment]
> > How do you plan to manage user sessions?
> 
> I don't think we need the concept of a session here. The session is the SMS thread. The context is maintained using UUID's in the request URLs. We can do client side sessions, but may have to brainstorm tradeoffs a bit more. My intention was to minimize UI requirements as much as possible.

You do that by hiding complexity in the client side code. The trade offs will be too great. It would be easier to validate a phone number with SMS verification to allow them access to the page(s).

I agree to using uuid in place of PII in the URL.

> > > Everything can be authenticated by including the phone number and a checksum in the URL sent to the user.
> > 
> > 
> > So the phone number and check sum match and they are granted access?
> 
> Yes iff the checksum matches the expected value from the server, then the submission is valid.
> 
> > What is some else copies that URL or shares it? If the check some expires 🧂 what if the actual person using the web app with that number loses internet connection? Or refreshes the page?
> 
> I think URL copies, shares and refreshes are ok, because all we need to do is perform uniqueness validation prior to submitting the entry to the DB. That said there are some interesting considerations here now that I think about it, with regard to using Firestore in the client with the "session-less" SMS model. We may need to consider using the GCP cloud function for all DB writes. Thanks for bringing this up. Additional notes inline.

Could you explain uniqueness? Only the first submission counts??

> We may need to consider using the GCP cloud function for all DB writes.
🔥 store cloud functions? If so they are slow and put a lag on the UI. They are good for deferred actions. Like sending a password reset email.

**Comment by un7c0rn** [comment]
@bernardbaker Thanks for the notes and great discussion here. Proposed plan for moving forward:

Investigate GDPR and PII requirements for supported countries #16 
Draft privacy design doc #17 
Design user sign in and data wipeout model #18 
Create GCP API endpoint for client DB writes #19 

Please let me know if you have bandwidth to take ownership on any of these. Ideally I still think the sign-in requirements within #18 can be minimized by leveraging SMS as our primary communication channel and performing backend validation, but I also agree with your point about people attempting to game the system when reward is significant. We'll of course have an opportunity to review all of this once we have an end-to-end MVP and start to do more comprehensive pre-launch testing.

**Comment by un7c0rn** [comment]
> > > How do you plan to manage user sessions?
> > 
> > 
> > I don't think we need the concept of a session here. The session is the SMS thread. The context is maintained using UUID's in the request URLs. We can do client side sessions, but may have to brainstorm tradeoffs a bit more. My intention was to minimize UI requirements as much as possible.
> 
> You do that by hiding complexity in the client side code. The trade offs will be too great. It would be easier to validate a phone number with SMS verification to allow them access to the page(s).
> 
> I agree to using uuid in place of PII in the URL.
> 
> > > > Everything can be authenticated by including the phone number and a checksum in the URL sent to the user.
> > > 
> > > 
> > > So the phone number and check sum match and they are granted access?
> > 
> > 
> > Yes iff the checksum matches the expected value from the server, then the submission is valid.
> > > What is some else copies that URL or shares it? If the check some expires 🧂 what if the actual person using the web app with that number loses internet connection? Or refreshes the page?
> > 
> > 
> > I think URL copies, shares and refreshes are ok, because all we need to do is perform uniqueness validation prior to submitting the entry to the DB. That said there are some interesting considerations here now that I think about it, with regard to using Firestore in the client with the "session-less" SMS model. We may need to consider using the GCP cloud function for all DB writes. Thanks for bringing this up. Additional notes inline.
> 
> Could you explain uniqueness? Only the first submission counts??
> 

Only a single submission for a given (challenge identifier, player identifier) pair, counts. Users may not submit to the same challenge twice.

> > We may need to consider using the GCP cloud function for all DB writes.
> > 🔥 store cloud functions? If so they are slow and put a lag on the UI. They are good for deferred actions. Like sending a password reset email.

I don't think the speed would materially impact the UI. We wouldn't need to perform a POST, we'd still use AJAX to communicate with the Cloud function. I imagine this would be less than a fraction of a second for most users and in fact the Firestore JS API is likely doing the same thing under the hood. The value add here is that we can perform all validation server side with no risk of a malicious user bypassing validation locally.

**Comment by bernardbaker** [comment]
> > > in the request for navigating to the challenge submission page matches the checksum sent to the user from the backend
> > 
> > 
> > So the game ID has a related checksum? And that game ID is extracted and verifed. In the client side code?
> 
> Actually, no. You raise a great point and I don't want to perform any client side validation in JS. We have a few options here:
> 
> 1. Implement a session based model using Firestore auth. This comes at the costs of users needing to go through an additional signup flow step, and from my perspective still doesn't solve the malicious user problem since one can simply hijack the JS session in their browser to perform direct writes to the DB.
> 2. Move all DB writes to GCP. We can simply extend the Python Firestore abstraction layer we have now and put the DB behind a Python cloud function. This way all writes **must** pass through server side validation for checksums and uniqueness.

> hijack the JS session in their browser to perform direct writes to the DB.
So a logged in user is doing this? Surely you'd just lock them out and disqualify them for tampering with the URL.

2) sounds good. So there will be a client side facade which calls the 🔥 store 🐍 API on GCP?

**Comment by un7c0rn** [comment]
> > > > in the request for navigating to the challenge submission page matches the checksum sent to the user from the backend
> > > 
> > > 
> > > So the game ID has a related checksum? And that game ID is extracted and verifed. In the client side code?
> > 
> > 
> > Actually, no. You raise a great point and I don't want to perform any client side validation in JS. We have a few options here:
> > 
> > 1. Implement a session based model using Firestore auth. This comes at the costs of users needing to go through an additional signup flow step, and from my perspective still doesn't solve the malicious user problem since one can simply hijack the JS session in their browser to perform direct writes to the DB.
> > 2. Move all DB writes to GCP. We can simply extend the Python Firestore abstraction layer we have now and put the DB behind a Python cloud function. This way all writes **must** pass through server side validation for checksums and uniqueness.
> 
> > hijack the JS session in their browser to perform direct writes to the DB.
> > So a logged in user is doing this? Surely you'd just lock them out and disqualify them for tampering with the URL.
> 
> 1. sounds good. So there will be a client side facade which calls the 🔥 store 🐍 API on GCP?

Exactly.

**Comment by bernardbaker** [comment]
> > > > How do you plan to manage user sessions?
> > > 
> > > 
> > > I don't think we need the concept of a session here. The session is the SMS thread. The context is maintained using UUID's in the request URLs. We can do client side sessions, but may have to brainstorm tradeoffs a bit more. My intention was to minimize UI requirements as much as possible.
> > 
> > 
> > You do that by hiding complexity in the client side code. The trade offs will be too great. It would be easier to validate a phone number with SMS verification to allow them access to the page(s).
> > I agree to using uuid in place of PII in the URL.
> > > > > Everything can be authenticated by including the phone number and a checksum in the URL sent to the user.
> > > > 
> > > > 
> > > > So the phone number and check sum match and they are granted access?
> > > 
> > > 
> > > Yes iff the checksum matches the expected value from the server, then the submission is valid.
> > > > What is some else copies that URL or shares it? If the check some expires 🧂 what if the actual person using the web app with that number loses internet connection? Or refreshes the page?
> > > 
> > > 
> > > I think URL copies, shares and refreshes are ok, because all we need to do is perform uniqueness validation prior to submitting the entry to the DB. That said there are some interesting considerations here now that I think about it, with regard to using Firestore in the client with the "session-less" SMS model. We may need to consider using the GCP cloud function for all DB writes. Thanks for bringing this up. Additional notes inline.
> > 
> > 
> > Could you explain uniqueness? Only the first submission counts??
> 
> Only a single submission for a given (challenge identifier, player identifier) pair, counts. Users may not submit to the same challenge twice.
> 
> > > We may need to consider using the GCP cloud function for all DB writes.
> > > fire store cloud functions? If so they are slow and put a lag on the UI. They are good for deferred actions. Like sending a password reset email.
> 
> I don't think the speed would materially impact the UI. We wouldn't need to perform a POST, we'd still use AJAX to communicate with the Cloud function. I imagine this would be less than a fraction of a second for most users and in fact the Firestore JS API is likely doing the same thing under the hood. The value add here is that we can perform all validation server side with no risk of a malicious user bypassing validation locally.

Ok. But check if they need to warm up first.  If you have the UI waiting on a response. That will be a lag before the page changes or a notification is shown. And if there is an error then they cant see a success message before a failure message.

**Comment by bernardbaker** [comment]
> > > I think a goal here is for people to be able to play this game without signing up for a new account or transitively granting access to any PII.
> > 
> > 
> > They won't be signing up for an account. It's a secure way to sign into the app. Verify a user session. Login. Logout.
> > Correct me if I'm wrong but the player enters their phone number. It's in the URL. This is PII.
> > My response to the aforementioned issues is...
> > I'm sure this can be managed, and free legal resources can be accessed through developer communities. The app does store phone numbers. The app should provide a sign in method. You can't rely on people playing by the rules, because it's a game.
> 
> Need more information on GDPR and PII here. I don't see how a sign in method changes things. In either case we agree that a phone number is PII, so we probably need a retention and wipeout policy. My question is does that policy require a user sign in. I'm not against implementing sign in if it's the right thing to do here. That said I think it's critical to get an end-to-end testable MVP in place before we draft our ToS. I also think per this discussion we cannot write to the Firestore DB from the client. Interested in your thoughts.
> 
> Thanks,
> Brandon

I agree with everything but the policy and sign in. The app needs a sign in. Not necessarily a sign up. The number will be stored when they sign in and a user is created in 🔥 store.

Because we store and process the number we have to address a privacy policy outlining how we manage and use a player's PII.

🛌 I'll catch up with you this evening GMT.

**Comment by un7c0rn** [comment]
@bernardbaker Agreed. Let's proceed with that plan.

Thanks!
Brandon

**Comment by bernardbaker** [comment]
@un7c0rn I haven't been able to do any work today. I was working on another project. I'll pick this up tomorrow morning.

**Comment by bernardbaker** [comment]
@un7c0rn created a PR https://github.com/un7c0rn/stopthevirus/pull/23

**Comment by bernardbaker** [comment]
Closing this issue 😺
